src/Cryptography/Sodium/Bindings/Hashing.hs

```diff
@@ -0,0 +1,178 @@
+-- Use the following constants as parameter to this function:
+--
+--   * 'cryptoGenericHashBytesMin' (16U)
+--   * 'cryptoGenericHashBytes' (32U)
+--   * 'cryptoGenericHashBytesMax' (64U)
+--
+-- @since 0.0.1.0
+allocateGenericHashStateOfSize :: CSize -> IO (Ptr CryptoGenericHashState)
+allocateGenericHashStateOfSize size = mallocBytes (fromIntegral size)
+
+-- | This function intialises a hash state with a key of a specified length, and will
+-- eventually produce an output with the specified length in bytes.
+--
+-- The @'Ptr' 'CUChar'@ argument must point to enough memory to hold a key,
+-- which must also be initialised.
+-- This is at least 'cryptoGenericHashKeyBytesMin', at most
+-- 'cryptoGenericHashKeyBytesMax', and should typically be 'cryptoGenericHashKeyBytes'.
+-- It is the caller's responsibility to ensure that these hold.
+--
+-- @since 0.0.1.0
+foreign import capi "sodium.h crypto_generichash_init"
+  cryptoGenericHashInit ::
+    -- | Pointer to the hash state
+    Ptr CryptoGenericHashState ->
+    -- | Pointer to a key
+    Ptr CUChar ->
+    -- | Length of the key
+    CSize ->
+    -- | Length of the result
+    CSize ->
+    IO CInt
+
+-- | If you process a message in chunks, you can sequentially process each chunk by calling 'cryptoGenericHashUpdate'
+-- by providing a pointer to the previously initialised state, a pointer to the input chunk,
+-- and the length of the chunk in bytes.
+--
+-- @since 0.0.1.0
+foreign import capi "sodium.h crypto_generichash_update"
+  cryptoGenericHashUpdate ::
+    -- | Pointer to the hash state
+    Ptr CryptoGenericHashState ->
+    -- | Pointer to a chunk to be processed
+    Ptr CUChar ->
+    -- | Length of the chunk in bytes
+    CULLong ->
+    IO CInt
+
+-- | After processing everything you need with 'cryptoGenericHashUpdate', you can finalise the operation
+-- with 'cryptoGenericHashFinal'.
+--
+-- @since 0.0.1.0
+foreign import capi "sodium.h crypto_generichash_final"
+  cryptoGenericHashFinal ::
+    -- | The hash state used throughout the previous hashing operations
+    Ptr CryptoGenericHashState ->
+    -- | The pointer to the resulting fingerprint
+    Ptr CUChar ->
+    CSize ->
+    IO CInt
+
+-- | Haskell binding to the @crypto_generichash_BYTES@ constant
+--
+-- It is a 32-bit unsigned integer (32U).
+--
+-- @since 0.0.1.0
+foreign import capi "sodium.h value crypto_generichash_BYTES"
+  cryptoGenericHashBytes :: CSize
+
+-- | Haskell binding to the @crypto_generichash_BYTES_MIN@ constant
+--
+-- It is a 16-bit unsigned integer (16U).
+--
+-- @since 0.0.1.0
+foreign import capi "sodium.h value crypto_generichash_BYTES_MIN"
+  cryptoGenericHashBytesMin :: CSize
+
+-- | Haskell binding to the @crypto_generichash_BYTES_MAX@ constant
+--
+-- It is a 64-bit unsigned integer (64U).
+--
+-- @since 0.0.1.0
+foreign import capi "sodium.h value crypto_generichash_BYTES_MAX"
+  cryptoGenericHashBytesMax :: CSize
+
+-- | Haskell binding to the @crypto_generichash_KEYBYTES@ constant
```

Comment:
`CSize` is not necessarily 32 bits wide. According to the C standard, it is _at least_ 32 bits wide, but could be larger. This information is both redundant (we have the type right there) and misleading.